File: JavaScript/Math-Array-String/65.js
// 977. Squares of a Sorted Array

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquares = function(nums) {
    let left = 0;
    let right = nums.length - 1;
    let result = new Array(nums.length);
    let index = nums.length - 1; // Start filling the result array from the end

    while (left <= right) {
        let leftSquare = nums[left] * nums[left];
        let rightSquare = nums[right] * nums[right];

        if (leftSquare > rightSquare) {
            result[index] = leftSquare;
            left++;
        } else {
            result[index] = rightSquare;
            right--;
        }
        index--;
    }

    return result;
};

console.log(sortedSquares([-4,-1,0,3,10])); // Output : [0, 1, 9, 16, 100]

File: Python/Math-Array-String/65.py
# 977. Squares of a Sorted Array

from typing import List

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        # Initialize two pointers and a result array
        left, right = 0, len(nums) - 1
        result = [0] * len(nums)
        index = len(nums) - 1  # Position to place the largest square

        while left <= right:
            left_square = nums[left] ** 2
            right_square = nums[right] ** 2
            
            if left_square > right_square:
                result[index] = left_square
                left += 1
            else:
                result[index] = right_square
                right -= 1
            
            index -= 1

        return result


sol = Solution()
output = sol.sortedSquares([-4,-1,0,3,10])
print(output) # # Output: [0, 1, 9, 16, 100]